velox/functions/prestosql/DecimalArithmetic.cpp

```diff
@@ -0,0 +1,205 @@
+ */
+
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+template <
+    typename R /* Result Type */,
+    typename A /* Argument1 */,
+    typename B /* Argument2 */,
+    typename Operation /* Arithmetic operation */>
+class DecimalBaseFunction : public exec::VectorFunction {
+ public:
+  DecimalBaseFunction(uint8_t aRescale, uint8_t bRescale)
+      : aRescale_(aRescale), bRescale_(bRescale) {}
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    if (args[0]->isConstantEncoding() && args[1]->isFlatEncoding()) {
+      // Fast path for (const, flat).
+      auto constant = args[0]->asUnchecked<SimpleVector<A>>()->valueAt(0);
+      auto flatValues = args[1]->asUnchecked<FlatVector<B>>();
+      auto rawValues = flatValues->mutableRawValues();
+      auto rawResults = prepareResults(rows, resultType, context, result);
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], constant, rawValues[row], aRescale_, bRescale_);
+      });
+    } else if (args[0]->isFlatEncoding() && args[1]->isConstantEncoding()) {
+      auto flatValues = args[0]->asUnchecked<FlatVector<A>>();
+      auto constant = args[1]->asUnchecked<SimpleVector<B>>()->valueAt(0);
+      auto rawValues = flatValues->mutableRawValues();
+      auto rawResults = prepareResults(rows, resultType, context, result);
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], rawValues[row], constant, aRescale_, bRescale_);
+      });
+    } else if (args[0]->isFlatEncoding() && args[1]->isFlatEncoding()) {
+      auto flatA = args[0]->asUnchecked<FlatVector<A>>();
+      auto rawA = flatA->mutableRawValues();
+      auto flatB = args[1]->asUnchecked<FlatVector<B>>();
+      auto rawB = flatB->mutableRawValues();
+      auto rawResults = prepareResults(rows, resultType, context, result);
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], rawA[row], rawB[row], aRescale_, bRescale_);
+      });
+    } else {
+      exec::DecodedArgs decodedArgs(rows, args, context);
+      auto a = decodedArgs.at(0);
+      auto b = decodedArgs.at(1);
+      BaseVector::ensureWritable(rows, resultType, context->pool(), result);
+      auto rawResults =
+          (*result)->asUnchecked<FlatVector<R>>()->mutableRawValues();
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row],
+            a->valueAt<A>(row),
+            b->valueAt<B>(row),
+            aRescale_,
+            bRescale_);
+      });
+    }
+  }
+
+ private:
+  R* prepareResults(
```

Comment:
Looks like all 4 if-else cases need this. Can we remove this call and initialize the result vector once before the if-else block?